packages/flutter/lib/src/services/platform_protocols.dart

```diff
@@ -0,0 +1,219 @@
+  /// to receive an error event in that case.
+  ///
+  /// The native implementation must wrap each data event in an envelope of the
+  /// form
+  ///
+  ///     { "status": "ok",
+  ///       "data": <data>
+  ///     }
+  ///
+  /// Similarly, an error event must be wrapped in an envelope of the form
+  ///
+  ///     { "status": "error",
+  ///       "message": <human-readable error message>,
+  ///       "data": <error details>
+  ///     }
+  ///
+  /// For each successful event, the specified [decoder] is used to turn the
+  /// data into a [T] instance (or `null`), which is then produced as a data
+  /// event by the returned stream. A `null` or missing decoder is treated
+  /// as a constantly `null` function.
+  ///
+  /// On each native error event, the returned stream will produce an error
+  /// event with a [PlatformException] containing the error message and error
+  /// details data.
+  ///
+  /// Malformed JSON or envelopes, or data that the [decoder] cannot handle,
+  /// will make the returned Stream produce an error event containing a
+  /// [FormatException].
+  static Stream<T> createJSONBroadcastStream<T>({
+    @required String channel,
+    Map<String, dynamic> args,
+    T decoder(dynamic json),
+  }) {
+    assert(channel != null);
+    assert(args == null || args['eventChannel'] is String);
+    final String eventChannel = args == null ? channel : args['eventChannel'];
+    StreamController<T> controller;
+    controller = new StreamController<T>.broadcast(
+      onListen: () async {
+        PlatformMessages.setJSONMessageHandler(
+          eventChannel,
+          (dynamic reply) async {
+            try {
+              controller.add(_interpretJSON<T>(reply, decoder));
+            } catch (e) {
+              controller.addError(e);
+            }
+          }
+        );
+        try {
+          await invokeJSONFunction<Null>(channel: channel, name: 'listen', args: args);
+        }
+        catch (e) {
+          PlatformMessages.setJSONMessageHandler(eventChannel, null);
+          controller.addError(e);
+        }
+      }, onCancel: () async {
+        PlatformMessages.setJSONMessageHandler(eventChannel, null);
+        try {
+          await invokeJSONFunction<Null>(channel: channel, name: 'cancel', args: args);
+        }
+        catch (exception, stack) {
+          FlutterError.reportError(new FlutterErrorDetails(
+            exception: exception,
+            stack: stack,
+            library: 'services library',
+            context: 'during de-activating platform stream on channel $channel',
+          ));
+        }
+      }
+    );
+    return controller.stream;
+  }
+
+  static T _interpretJSON<T>(dynamic json, T decoder(dynamic json)) {
+    if (json is Map) {
+      final dynamic status = json['status'];
+      final dynamic message = json['message'];
+      final dynamic data = json['data'];
+      if (status == 'ok') {
```

Comment:
This one is part of a chain? I've rewritten the chain so that it now contains one-liners in all branches, and dispensed with the braces.